content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
++++
+title = "Framework to include runtime functions to your llvm pass"
+extra.author = "Sachille Atapattu"
+extra.author_link = "https://www.csl.cornell.edu/~sachille/"
+extra.bio = """
+  [Sachille][] is a third-year graduate student working across the stack to 
+  make heterogeneous computing easier. Sachille also likes books (currently reading 
+  Endure by Alex Hutchinson) and philosophizing with friends.
+
+[sachille]: https://www.csl.cornell.edu/~sachille
+"""
++++
+
+This blog is relating experience of the author to efficiently and productively 
+include runtime functions as an easy way to simulate hardware accelerators.
+
+My research direction, which is to make heterogeneous architectures accessible to 
+software programmers, requires implementing overall system solutions. This becomes
+surprisingly tricky as implementing hardware is hard, some (most?) heterogeneous 
+architectures have their own custom and close-sourced programming models and existing 
+popular simulators also require limiting to conventional programming models. So, to try
+out a new architecture idea, either you have to limit yourself to existing toolflow ([Dahlia][dahlia]),
+or extend simulators with the programming models they support ([ParallelXL][parallelxl]). 
+Me being a hardware programmer( and a newbie to the software end of the stack), 
+I also have aversion to start implementing a full-stack solution without sufficient 
+insight to justify I'm doing something sane.
+
+An easy hack I came across is using runtime functions to do my simulations.
+This way, I don't have to put my abstraction level in concrete, as it's merely 
+creating a new llvm pass at a different level with the accompanying python simulator, 
+(which though non trivial, is still better than creating an extension for [gem5][gem5].)
+
+In summary this report is to say, this is a great hack to gain intuition, however 
+to make a truly general simulator framework may need a lot more engineering. And of 
+course, running python functions is super slow.
+
+[dahlia]: https://drive.google.com/file/d/13zJDMHOkymitZc6Zu4U7QXkRFSr99Lk_/view?usp=sharing 
+[parallelxl]: https://www.csl.cornell.edu/~tchen/files/parallelxl-micro18.pdf
+[gem5]: https://www.gem5.org
+
+## Where to begin
+[llvm for grad students][grad] provides a basic skeleton to use runtime functions through llvm.
+This pass simply inserts a logging function in a C runtime library after an instruction. [Skeleton][skeleton]
+provides a starting point for this framework, which I'm calling Leech. [Leech][leech] can be found here.
+
+[grad]: https://www.cs.cornell.edu/~asampson/blog/llvm.html
+[skeleton]: https://github.com/sampsyo/llvm-pass-skeleton
+[leech]: https://github.com/sa2257/llvm-runtime-pass
+
+## What this framework is targeting
+I'm defining some key levels to insert a runtime function replacements.
+- Instruction level
+- Subgraph level
```

Comment:
Define this? The other things are basic LLVM constructs, but this needs more explanation.